brave/brave-core
# (Wallet) Implement Bitcoin support and network selection during onboarding flow

## Resolves https://github.com/brave/brave-browser/issues/32721
## Resolves https://github.com/brave/brave-browser/issues/36578

🎨 Figma design: https://www.figma.com/file/oXIiv45pIhF2RGlpEjOfUg/Wallet-onboarding?node-id=6223%3A70942&mode=dev

## UI Preview

|Light Mode |Dark Mode|
|-----|--------|
|![Screenshot 2024-04-08 at 3 19 56 PM](https://github.com/brave/brave-core/assets/3308503/a4db14d3-acdc-4681-8b62-185bde752645)|![Screenshot 2024-04-08 at 3 20 38 PM](https://github.com/brave/brave-core/assets/3308503/97e01eb5-bcdd-4813-bb7d-ff8191ee090a)|

## Demo


https://github.com/brave/brave-core/assets/3308503/70d39c8e-c522-4cd4-b429-50a4e5f6b63a



## Details

* Enable Bitcoin support by default on Android
* Return Bitcoin logo for Bitcoin mainnet and testnet
* Return Neon logo for Neon network 
* Remove unused methods from `CryptoModel`, `NetworkModel`, `NetworkSelectorModel`, and `NetworkSelectorActivity`
* Perform refactoring to onboarding fragments
* Return keyring IDs for Bitcoin mainnet and testnet
* ⚠️ Overhaul viewpager layout, simplify Android layout pass and [fix `notifyDataSetChanged` issue](https://issuetracker.google.com/issues/186366027)
* Implement new `AutoGridLayoutManager` that adapts its columns to recycler view size
* Implement new `OnboardingNetworkSelectionFragment` following Figma design

This PR introduces Bitcoin support for Brave Wallet on Android and implements a new screen that let the user to choose which networks to display. Bitcoin and Filecoin accounts wille be created *only* when their network is selected in the network selection screen. In the network selection screen all the test networks are not selected by default, hence not displayed once the Wallet gets created. Ethereum and Solana accounts instead are pre-selected by default and cannot be deselected. Unselected networks will be added to the hidden network list, accessible in the Wallet settings section (to be implemented soon).

## Follow up Issues

- Add 500 ms delay on Wallet creation screen to avoid flickering - https://github.com/brave/brave-browser/issues/37395
- Refactor fragment creation adapter for onboarding flow - https://github.com/brave/brave-browser/issues/37396
- Implement logic to show hidden network list in the Wallet settings section - https://github.com/brave/brave-browser/issues/37397
- Remove unused code (check also code unused by removed + button) - https://github.com/brave/brave-browser/issues/37398
- Remove live data data type where uneeded - https://github.com/brave/brave-browser/issues/37399
- Fix new network names following Dekstop implementation - https://github.com/brave/brave-browser/issues/37400


## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

### Bitcoin Support Test Plan

- Reset and clear wallet data
- Open Wallet onboarding flow
- On network selection screen, select Bitcoin network and tap continue
- Finish onboarding process
- Navigate to account section
- Observe Bitcoin account is present and is called "Bitcoin Account 1"

### Network Selection Screen Test Plan

- Reset and clear wallet data
- Open Wallet onboarding flow
- On network selection screen, tap on "Show testnets"
- Observe test networks are shown
- Try to deselect Ethereum network
- Observe Ehereum network cannot be deselected
- Try to deselect Solana network
- Observe Solana network cannot be deselected
- Select Bitcoin account
- Observe the button reflects the number of selected networks
- Tap on "Select all" button
- Observe all popular networks get selected
- Tap on search bar
- Type "fil"
- Observe Filecoin networks are shown
- Tap on Filecoin mainnet
- Tap on continue button
- Finish onboarding process
- Observe Bitcoin account and Filecoin account are present
